Added `\yii\i18n\Formatter::asDurationTime()`

githubjeka opened the pull request:
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #12022 fixed by the PR |

githubjeka committed:
Added `\yii\i18n\Formatter::asDurationTime()`

SilverFire commented on framework/i18n/Formatter.php:
```diff
@@ -828,14 +827,66 @@ public function asRelativeTime($value, $referenceTime = null)
+     */
+    public function asDurationTime($value, $options = [])
+    {
+        if (!isset($options['implodeString'])) {
```
H

SilverFire replied:
How about

```php
$implodeString = ArrayHelper::get($options, 'implodeString', ', ');
```

SilverFire commented on framework/i18n/Formatter.php:
```diff
@@ -849,37 +900,53 @@ public function asDuration($value, $implodeString = ', ', $negativeSign = '-')
 
         if ($interval->y > 0) {
-            $parts[] = Yii::t('yii', '{delta, plural, =1{1 year} other{# years}}', ['delta' => $interval->y], $this->locale);
+            $parts[] = Yii::t('yii', '{delta, plural, =1{1 year} other{# years}}', ['delta' => $interval->y],
```
Splitting for two lines does not look reasonable here

githubjeka replied:
Fixed

SilverFire reviewed (changes requested):
And we still have no way to configure years, months and days

samdark commented on framework/i18n/Formatter.php:
```diff
@@ -828,14 +827,66 @@ public function asRelativeTime($value, $referenceTime = null)
      * @param string $negativeSign will be prefixed to the formatted duration, when it is negative. Defaults to `-`.
      * @return string the formatted duration.
      * @since 2.0.7
+     * @deprecated Since version in 2.0.10. The signature will be changed in 2.1 to [[Formatter::asDurationTime()]]
```
`in` isn't needed here. Just `Since version 2.0.10`

githubjeka replied:
fixed thanks

SilverFire commented:
Maybe, we should agree three different options for every part:

- `null` (default) - display part only when it's not zero
- `true` - always display a part, even it's zero
- `false` - never display a part

githubjeka commented:
is label `2.0.11` ?

SilverFire commented:
The PR does not look ready to be merged right now, so I've assigned it to the next milestone.
If you can continue work on this PR, we can accept it in 2.0.10 once it's ready

githubjeka committed:
fix typos

githubjeka committed:
Refactoring to Formatter::asDuration

Added more tests.
Added config to years, month, days.
Added handler for TRUE value.
Update phpDoc

githubjeka commented:
Corrected

samdark commented on framework/i18n/Formatter.php:
```diff
@@ -849,37 +906,44 @@ public function asDuration($value, $implodeString = ', ', $negativeSign = '-')
         }
 
-        if ($interval->y > 0) {
+        if ($showYears || (is_null($showYears) &&$interval->y > 0)) {
```
Formatting! `&&$inteval`. Same in more lines ↓

samdark replied:
`=== null` should be used instead of `is_null`

samdark commented on framework/i18n/Formatter.php:
```diff
@@ -828,18 +828,75 @@ public function asRelativeTime($value, $referenceTime = null)
+     */
+    public function asDurationTime($value, $options = [])
+    {
+        if (is_null($value)) {
```
`=== null` should be used instead of `is_null`

githubjeka replied:
Why? Is requirement to yii2 code?

samdark replied:
It's overall better to use simple checks instead of function calls where possible. More explicit, less opcodes.

SilverFire commented:
Btw, the implementation is not very smart.

Take a look at implementation in `moment.js` https://github.com/moment/moment/blob/develop/src/lib/duration/humanize.js

githubjeka committed:
is_null to simple checks & formatter for code

githubjeka commented:
@SilverFire I think about. But it will pull the more changes. And so, I added a few lines :) May be better create new issue...

SilverFire commented:
It's fine to do it in this pull request.

SilverFire commented:
Considering that we implement completely different behavior, I suggest to leave `asDuration()` almost changes and implement `asRoudedDuration()` that will create `DateInterval` object and round it so only required parameters are left

githubjeka commented:
It will called DecoratorDateInterval class, formatter should use it.... #эко_меня_понесло

githubjeka committed:
Added more tests

githubjeka commented:
>Does it make any sense to show something like 1 year 11 months 3 seconds ago?

Yes you can - via `\yii\i18n\Formatter::asRelativeTime`. It isn't about this issue.

I added more tests. Now, I suggest leave as is or please explain what do you want to exactly?

cebe commented:
Sorry for the late reply. It seems that the requirements for this method are not very clear so before putting more work into this we should agree on what the different methods should do. Also the interface for specifying which part of the duration to show is not simple. It ususally does not make sense to display e.g. only year and seconds so you display `1 year, 16 seconds` for a timestamp that was actually a year, a month and some hours + 16 seconds. What should instead applied is a parameter that defines the precision. There are also multiple options on how to handle that:

- precision is `year`:
  - we could display `0 years` if duration <= 1 year, `1 year` if <= 2 year etc.
  - alternative: display `x sec` if duration <= 1 min, `x min` if duration <= 1 h, and switch to above display if we reached year
- precision is `month`:
  - `1 year, 2 month` if > duration is over a year or just display the months if it is not.
  - same couting as above could be applied. (or not)

There are a lot of options here that should be discussed to define a method interface (or multiple methods).
See also 
- http://www.icu-project.org/apiref/icu4j/com/ibm/icu/text/MeasureFormat.html
- http://www.icu-project.org/apiref/icu4j/com/ibm/icu/text/RelativeDateTimeFormatter.html
guess intl has something we could use here. Need to find it.
